app/services/chatbot/chatbot.ts

```diff
@@ -0,0 +1,203 @@
+    return new Promise((resolve, reject) => {
+      const url = this.apiEndpoint('login');
+      const headers = authorizedHeaders(this.userService.apiToken);
+      headers.append('Content-Type', 'application/json');
+      const request = new Request(url, {
+        headers,
+        method: 'POST',
+        body: JSON.stringify({})
+      });
+
+      fetch(request)
+        .then(handleErrors)
+        .then(response => response.json())
+        .then((response: ChatbotApiServiceState) => {
+          this.LOGIN(response);
+          resolve(true);
+        })
+        .catch(err => {
+          reject(err);
+        });
+    });
+  }
+
+  //
+  // GET requests
+  //
+
+  fetchDefaultCommands() {
+    const url = this.apiEndpoint('commands/default', true);
+    const headers = authorizedHeaders(this.state.api_token);
+    const request = new Request(url, { headers });
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json());
+  }
+
+  fetchCustomCommands(page: number) {
+    const url = this.apiEndpoint(`commands?page=${page}`, true);
+    const headers = authorizedHeaders(this.state.api_token);
+    const request = new Request(url, { headers });
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json());
+  }
+
+  fetchTimers(page: number) {
+    const url = this.apiEndpoint(`timers?page=${page}`, true);
+    const headers = authorizedHeaders(this.state.api_token);
+    const request = new Request(url, { headers });
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json());
+  }
+
+  //
+  // POST, PUT requests
+  //
+  updateDefaultCommand(slugName: string, commandName: string, data: DefaultCommand) {
+    const url = this.apiEndpoint(`settings/${slugName}/commands/${commandName}`, true);
+    const headers = authorizedHeaders(this.state.api_token);
+    headers.append('Content-Type', 'application/json');
+    const request = new Request(url, {
+      headers,
+      method: 'POST',
+      body: JSON.stringify(data)
+    });
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json());
+  }
+
+  updateCustomCommand(id: string, data: CustomCommand) {
+    const url = this.apiEndpoint(`commands/${id}`, true);
+    const headers = authorizedHeaders(this.state.api_token);
+    headers.append('Content-Type', 'application/json');
+    const request = new Request(url, {
+      headers,
+      method: 'PUT',
+      body: JSON.stringify(data)
+    });
+
+    return fetch(request)
+      .then(handleErrors)
+      .then(response => response.json());
+  }
+
+  createCustomCommand(data: CustomCommand) {
+    const url = this.apiEndpoint('commands', true);
+    const headers = authorizedHeaders(this.state.api_token);
+    headers.append('Content-Type', 'application/json');
+    const request = new Request(url, {
+      headers,
+      method: 'POST',
+      body: JSON.stringify(data)
+    });
```

Comment:
Oh, we definitely should simplify service requests in our project. I have a `request` function that I use in `WidgetSettingsService` but looks like I have to expose it for all parts of code like that.

Reply:
im simplified it further already to have `this.api(method, url, body)` instead of duplicate code, but your widget-settings is quite comprehensive. Just need to expose it and handle diff tokens